tests/sparseml/core/lifecycle/test_event.py

```diff
@@ -0,0 +1,565 @@
+        assert events == []
+        assert func_name == expected_func_name
+
+    def test_events_from_type_raises_value_error(self):
+        lifecycle = _get_event_lifecycle()
+        with pytest.raises(ValueError):
+            lifecycle.events_from_type("invalid")
+
+    @pytest.mark.parametrize(
+        "kwargs, increment ,expected",
+        [
+            ({"batches_per_step": None}, False, True),
+            ({"batches_per_step": 1}, False, True),
+            ({"batches_per_step": 3, "batch_count": 5}, False, True),
+            ({"batches_per_step": 4, "batch_count": 7}, False, True),
+            ({"batches_per_step": 4, "batch_count": 9}, False, False),
+            ({"batches_per_step": 4, "batch_count": 9}, True, False),
+            ({"batches_per_step": 4, "batch_count": 11}, True, True),
+        ],
+    )
+    def test_check_step_batches_count(self, kwargs, increment, expected):
+        lifecycle = _get_event_lifecycle()
+
+        for key, value in kwargs.items():
+            setattr(lifecycle, key, value)
+
+        actual = lifecycle.check_step_batches_count(increment=increment)
+
+        if increment:
+            if not expected:
+                assert lifecycle.batch_count == kwargs["batch_count"] + 1
+            else:
+                assert lifecycle.batch_count == 0
+
+        assert actual == expected
+
+    @pytest.mark.parametrize(
+        "kwargs, increment ,expected",
+        [
+            ({"invocations_per_step": None}, False, True),
+            ({"invocations_per_step": 1}, False, True),
+            ({"invocations_per_step": 3, "step_count": 5}, False, True),
+            ({"invocations_per_step": 4, "step_count": 7}, False, True),
+            ({"invocations_per_step": 4, "step_count": 9}, False, False),
+            ({"invocations_per_step": 4, "step_count": 9}, True, False),
+            ({"invocations_per_step": 4, "step_count": 11}, True, True),
+        ],
+    )
+    def test_check_step_invocations_count(self, kwargs, increment, expected):
+        lifecycle = _get_event_lifecycle()
+
+        for key, value in kwargs.items():
+            setattr(lifecycle, key, value)
+
+        actual = lifecycle.check_step_invocations_count(increment=increment)
+
+        if increment:
+            if not expected:
+                assert lifecycle.step_count == kwargs["step_count"] + 1
+            else:
+                assert lifecycle.step_count == 0
+        assert actual == expected
+
+
+class TestWrappedOptimEventLifecycle:
+    @pytest.mark.parametrize(
+        "method_name",
+        [
+            "batch_start_events",
+            "batch_end_events",
+        ],
+    )
+    def test_wrapped_lifecycle_raises_value_error_on_erroneous_invocation(
+        self, method_name
+    ):
+        lifecycle = _get_event_lifecycle(lifecycle_class=WrappedOptimEventLifecycle)
+        with pytest.raises(ValueError, match="batch"):
+            method = getattr(lifecycle, method_name)
+            method()
```

Comment:
Explanation on why this should fail would be useful

Reply:
Done!